### ISSUE TITLE
Dependency injection in Orleans Grains

### ISSUE THREAD

**Issue by nehmebilal** [dotnet/orleans#669]
In #443, we made a first attempt to allow dependency injection (DI) in Orleans. Even though we added a new constructor that allow injecting the `IGrainRuntime` and `IGrainIdentity`, this constructor was purely added for testing and is not really what DI is all about. Proper DI would allow one to inject any type into an Orleans grain in a generic way.

For instance, if my grain uses a service, let's say `IWeatherService`, I should be able to inject that type into an Orleans grain and use it without the need to know how the corresponding instance was created. This will **decouple the grain from the implementation** of `IWeatherService` and from **managing it's lifetime** (creation, singleton or not, etc). 

Today, to use `IWeatherService` in my grain, I have to do something like:
```csharp
public override Task OnActivateAsync()
{
    _weatherService = new WeatherServiceImplementation();
}
```

Ideally, I should be able to configure Orleans runtime at startup by registering an instance of `IWeatherService` that would then be automatically passed to the grain by the runtime on activation.

One simple way to achieve this is to add a dependency resolver parameter to `OnActivateAsync` as follows:

```csharp
public override Task OnActivateAsync(IDependencyResolver dependencyResolver)
{
    _weatherService = dependencyResolver.Resolve<IWeatherService>();
}
```

This approach allow one to replace the implementation of `IWeatherService` with another implementation (or with a test double) and to use a **DI container** for wiring the application (inversion of control, object creation, the good singleton, lazy initialization and all other goodies that a DI container offers). 

A better alternative would be for the grain to not deal directly with the `IDependencyResolver` but to directly specify what services are needed as parameters of the constructor or `OnActivateAsync` as follows:

```csharp
public Task OnActivateAsync(IWeatherService weatherService, IEmailService emailService)
{
    _weatherService = weatherService;
    _emailService = emailService;
}
```
In this case, the Orleans runtime will dynamically discover the parameters of `OnActivateAsync` (or constructor) and pass-in the appropriate implementation when `OnActivateAsync` is called. The runtime uses the `IDependencyResolver` internally to resolve the implementations.

It's worth noting that it doesn't stop at injecting services, one can also inject a grain into another grain:
```csharp
public Task OnActivateAsync(IWeatherService weatherService, IHelloGrain helloGrain)
{
...
}
```

Such support for dependency injection would be a great addition to Orleans. 

Please take a look at how ASP.NET provides support for DI [here](http://www.asp.net/web-api/overview/advanced/dependency-injection); it's a very similar use case to Orleans.

Thoughts?

**Comment by Plasma** [comment]
Great. My preference is injecting via the constructor because that's how other services and MVC controllers use DI (so this is normal, at least for me), and we normally place dependencies at the top of the class definition above the constructor and (trivially worth noting) ReSharper always offers "Initialize from constructor" as an option. We can also mark them readonly.

For this to make sense, I assume we construct a grain and call OnActivateAsync once, and when its time to be deactivated we call OnDeactivateAsync and then never use that Grain (C# object instance) again?

Cheers

**Comment by nehmebilal** [comment]
I agree that constructor would be the preferred choice but I am not sure if it's possible. 
In fact, Orleans runtime first creates a `Grain` using the default constructor and then configures it (sets the `GrainIdentity`, `GrainFactory`, etc). If we do constructor injection, we need to keep in mind that the `Grain` is not active until `OnActivateAsync` is called. In other words, you can't do certain operations in the constructor. For this reason, injection through `OnActivateAsync` seemed like the easier choice but Orleans team are probably in a better position to comment on this.

> For this to make sense, I assume we construct a grain and call OnActivateAsync once, and when its time to be deactivated we call OnDeactivateAsync and then never use that Grain (C# object instance) again?

Which is the case today isn't? A new object instance is constructor for each activation?

The trick with Orleans is that grains are distributed and can be moved from one silo to another. As a result, the injected dependencies must be stateless (unless the grain implementation itself is taking care or restoring their state inside `OnActivateAsync`) and they need to be reconstructed every time the grain is activated (except when the grain is reactivated on the same silo); all of which should be configurable so the user can choose things such as: always create a new instance of a dependency on activation or reuse existing instance when possible (same silo), etc.

**Comment by gabikliot** [comment]
Would not a better approach be to allow access to `IDependencyResolver`  via some kind of a new provider? Lets say we add a new provider, `IDependencyResolverProvider` that will be accessible to grain code via ` IGrainRuntime `.

That way we will preserve backward compatibility, which is becoming an increasingly important issue (#646) and will also not increase the mental load on the first time user, who needs to understand lots of new concepts and throwing `IDependencyResolver` into his face as part of the first public API he looks at (`OnActivateAsync`) really complicates the mental load, don't you think? Is dependency injection something we want the developer to start from, be among the first things about Orleans APIs he sees?

**Comment by nehmebilal** [comment]
> @gabikliot Gabi Would not a better approach be to allow access to IDependencyResolver via some kind of a new provider? Lets say we add a new provider, IDependencyResolverProvider that will be accessible to grain code via IGrainRuntime.

That is definitely an option and has the advantage of being backward compatible. However, if the constructor option is feasible (which is also backward compatible), that would be better.

> Is dependency injection something we want the developer to start from, be among the first things about Orleans APIs he sees?

DI is becoming the common way of wiring an application and I suspect that in the future, developers wouldn't even know how to use a framework without it.

Actually, wouldn't the following be backwards compatible? (if you have no parameters it simply means you have no dependencies to inject)
```csharp
public Task OnActivateAsync(IWeatherService weatherService, IEmailService emailService)
{
    _weatherService = weatherService;
    _emailService = emailService;
}
```

**Comment by yevhen** [comment]
> DI is becoming the common way of wiring an application and I suspect that in the future, developers wouldn't even know how to use a framework without it.

gosh, what a desperate future!

**Comment by yevhen** [comment]
@nehmebilal 

Do you realize that DI **!=** container?

**Comment by nehmebilal** [comment]
@yevhen It amazes me how friendly you always sound :smile:  

What I am suggesting here doesn't require the use of a DI container; but designing for DI opens the door to use a DI container.

**Comment by yevhen** [comment]
@nehmebilal believe me, I'm very friendly. You just don't know me :smile:

Right. What I suggest (and already suggested before) is just give a people a way to construct an instance of grain. They will figure out then, how to plug their own favorite container if needed.

Most of the time there will be only a few actors and so using simple conditional construction like `if (type == typeof(TActor)) return new TActor(dep1, dep2) ...` inside of the Composition Root is all what is needed, without incurring the overhead and performance loss of container-based construction.

**Comment by nehmebilal** [comment]
> believe me, I'm very friendly. You just don't know me :smile:

:+1: 

> What I suggest (and already suggested before) is just give a people a way to construct an instance of grain. They will figure out then, how to plug their own favorite container if needed.

Right now, Orleans runtime is the one that creates actors and configure them. How would you allow people to construct a grain?

> Most of the time there will be only a few actors and so using simple conditional construction like if (type == typeof(TActor)) return new TActor(dep1, dep2) ... inside of the Composition Root is all what is needed, without incurring the overhead and performance loss of container-based construction.

Hmm I don't know about that.. Dependencies are not only grains but other services. In addition, an Orleans based application can be part of a bigger application that can get as big as it wants. 

Regardless, whether you use a DI container or not, you need a way to inject dependencies into a grain, either by constructing the grain yourself (which is not possible currently) or by having a back door into the runtime that allow you to pass-in those dependencies (much like how it's done in ASP.Net).

**Comment by nehmebilal** [comment]
Okay it sounds like we have few options:

1- Add `IDependencyResolver` (or `IDependencyProvider`) to `IGrainRuntime`.

2- Add the dependencies directly to the parameters of `OnActivateAsync` and let the runtime discover the needed dependencies for a given grain and resolve them using the dependency resolver (user defined). This can happen at either compile or run time.

3- Add `IDependencyResolver` as a parameter to `OnActivateAsync`. This is simple but not backwards compatible.

4- Constructor injection. This could get very tricky because grains are created by the runtime and are not considered active until `OnActivateAsync` is called (there is no user code that is executed before `OnActivateAsync`). To use constructor injection, we will have to add some constraints on what can be done in the constructor which is error prone.

I personally agree with @gabikliot that 1- would be the simplest way to enable DI in Orleans and covers the needed use cases. I also agree with @Plasma that constructor injection is more natural for an OO programmer but considering that this is an actor programming framework, DI on activation can become the new natural :smiley: 

To resolve a dependency, the user code would look like the following:

```csharp
public override Task OnActivateAsync()
{
    _weatherService = Runtime.DependencyResolver.Resolve<IWeatherService>();;
}
```

To inject the dependencies, we can use a similar approach to what is used to provide [custom storage providers](http://dotnet.github.io/orleans/Step-by-step-Tutorials/Custom-Storage-Providers). In other words, the user will implement `IDependencyResolver` (where they can use a DI container) as a class library and point the runtime to it via configuration.

**Comment by gabikliot** [comment]
:+1: :+1:

**Comment by jason-bragg** [comment]
I’d prefer 1) as well.

As for DI implemented as a provider, I agree that it needs be a pluggable component, for integration purposes, with a default implementation provided by Orleans, but since DI/IOC is intended to support pluggable components, I suggest a better pattern would be to use the default DI solution to load a user provided DI implementation, if provided, than use that system to load all other pluggable parts of the framework, including providers.

I suggest the development of a minimal set of DI/IOC interfaces geared towards a small surface area.  A similarly minimalistic default implementation should also be developed.  Keeping the surface area small increases the likelihood it will support a subset of a large set of other, more complete, DI/IOC solutions.  Since a minimal DI interface can be used to acquire more expressive DI interfaces of more complete implementations, this minimal interface should not hinder application developers, but will reduce integration cost and learning curve.

I think we want to avoid trying to implement a complete DI/IOC system because Orleans probably does not need it, and others are focused on this effort.  Identifying the minimal set of capabilities we need, and keeping our usage scoped but extensible for the application layer would be my preference.

**Comment by Plasma** [comment]
I've been thinking about this a little more and traditionally I use DI like this:

1. For web-based requests (short lived), most services are either Singleton (they are thread safe across all requests), or per-request, so they get new()'d up for each new request and disposed of (not thread safe -- but is fine to re-use the same instance between other services used during that request). A web server may serve dozens to 100s of req/sec etc.

1. For job-processing requests (short to several minutes), it has say 4 worker threads on a 4 core machine and will scope dependencies for each job being processed. Singleton's are still shared across the 4 job threads, but "per request" dependencies are new()'d up for each job and disposed of at the end. So there may be say 4 jobs per machine at one time in progress.

1. Grains are a little different. A grain can live for a long time (hours), and I could have many of them (100,000+). 

If a Grain was injected with a "per grain" service then I would have 100,000x service object instantiates (and their dependencies) kept in memory; this is probably poor for performance.

Considering Orleans uses cooperative multitasking with (say 4) worker threads, perhaps another hybrid style in this case was to have service instances shared between Grains allocated to each worker (thus we re-use the same service instance across many grain instances, but only one of those grains is processing a message by that thread at one point in time). So we'd have 4 instances of the services for 100,000 grains in this example. However, this isn't how the Grain/Worker thread dynamic works right now, I assume the next free worker thread processes the next waiting message across all Grains.

Also, in the case of Web/Job processing, I have hook points on pre/post request/job start/end that I can use to start/stop a new DI scope that causes the resolving dependencies to be scoped correctly - I am not sure I have access to these points right now in Orleans (eg pre/post grain allocation, unless I make a new base grain class and override OnActivate/Deactivate).

A lot of these DI challenges may be solved if Orleans simply asked a static resolver to resolve the instance of a Grain (by default it uses its own resolver), where-as I could put SimpleInjector in the middle to resolve our dependencies instead perhaps.

**Comment by veikkoeeva** [comment]
@Plasma I remember @attilah had an idea and some rough scetches in Gitter some months ago supporting use cases you have. Unfortunatelly he's rather busy at the moment, as far as I know. With regards to [nightly Gitter] (as seen from here) for setting up DI, perhaps [these notes](https://github.com/dotnet/orleans/issues/626#issuecomment-126753794) by @nehmebilal give you a head start. It might be that designing a DI system for Orleans, one would need to share and sync the DI container state across silos (or reload the state in some other way).

By the way, @nehmebilal thanks for the example. It was a good one.

**Comment by veikkoeeva** [comment]
Oh, what @jason-bragg and @gabikliot wrote also. Having minimal infrastructure to load a general purpose DI container, if so desired, is the way to go.

**Comment by attilah** [comment]
I've a working pluggable implementation, which supports construction injection and also the resolution of grains from the DI container with 1 level of dependency scope (Silo level).

What is left is to implement the provider support, similar to what Orleans already has as Bootstrap providers, with a difference that only 1 DI provider can be configured for a Silo.

**Comment by nehmebilal** [comment]
I understand that there are many aspects to consider regarding the lifetime of the dependencies but I'd suggest we solve one problem at a time.

The current behaviour is to create the dependencies every time the grain is activated and let's not try to change that (right now) but focus on **moving the interface/implementation mapping of dependencies outside of the grain** (i.e. inversion of control) to improve testability and reduce coupling.

It is a simple as adding the `IDependencyResolver` to `IGrainRuntime` to allow the users to plug-in their favorite DI container (using the provider model). This should be simple and easy to achieve.

The lifetime of dependencies can be managed by the user (they can configure their DI container the way they want). I also don't see the need for a default DI container; it doesn't hurt but let's not make it a requirement (to reduce the scope of this issue).

We can start by defining the `IDependencyResolver` interface and then look at what is needed to wire it in the runtime.

**Comment by yevhen** [comment]
Are you guys kidding, right? Because it can't be true. Otherwise, I'm sleeping and this is a really bad dream.

DI is constructor/property injection and nothing else. The thing that you're proposing with `IDependencyResolver` is utter nonsense. It's a ServiceLocator pattern, but not a DI.

The point is: nobody really need `IDependencyResolver` property inside `IGrainRuntime`. Anyone can achieve the same with BootstrapProvider. Either we add support for normal ctor/property injection like what @attilah started or just don't bother at all.

Cheers!

**Comment by gabikliot** [comment]
How can you achieve this with BootstrapProvider? It does not expose any "resolve" functionality.

**Comment by attilah** [comment]
I'm implementing a new provider type. Just using Bootstrap provider as a "pattern".

**Comment by yevhen** [comment]
@gabikliot cmon, just create bootstrapovider which exposes `IContainerResolver` as static getter. Use its `Init` method to configure container. 5 minutes and you're done.

**Comment by ReubenBond** [comment]
I actually configure a static service locator in a Bootstrap provider as a hack until we have constructor/property injection.

I agree with @yevhen. We should not worry about lifecycles, at least not yet.

**Comment by nehmebilal** [comment]
The static service locator in a Bootstrap provider is a global state. Let's not go through the drawbacks of global state again :smiley: 

 In our case, the service locator is an interface that is passed to the constructor of the grain (not a global state) and can be replaced with a different implementation. In addition, all dependencies of the grain can be replaced with different implementations or test doubles which provides **inversion of control** (which is the main goal of DI).

On the other hand, I agree that if the grain explicitly lists it's dependencies as parameters of the constructor (or `OnActivateAsync`) that would benefit code readability and reduce boiler-plate code (`resolver.Resolve` calls won't be needed), but that's about it.

I don't think the use of constructor instead of `OnActivateAsync` has any advantages. In Orleans, an actor becomes active when `OnActivateAsync` is called, the constructor is pretty much not used. Property injection might be the better option:

```csharp
public class HelloGrain : Grain, IHelloGrain
{
    [Inject]
    public IWeatherService WeatherService;
}
```
Orleans can dynamically discover properties to be injected and set them before calling `OnActivateAsync`.

To sum up, adding `IDependencyResolver` to `IGrainRuntime` is definitely an improvement and offers *inversion of control*. Explicitly listing the dependencies as parameters of `OnActivateAsync` or using property injection offers the same benefits but improves code readability (but is harder to implement).

So should we go the easy way that offers inversion of control or the harder but cleaner way using property injection? :smiley: 

@attilah can you give more details about the approach you're proposing with some usage examples?

**Comment by Plasma** [comment]
I've prototyped the [Inject] usage (framework support at https://github.com/dotnet/orleans/compare/master...Plasma:feature-dependency-resolver) and I use the SimpleInjector ExecutionContextScopeLifestyle and create the context at the start/end of a grain message invocation, and inject/replace all [Inject] properties on the class:

````csharp
	[AttributeUsage(AttributeTargets.Property)]
	public class InjectAttribute : Attribute
	{
		
	}

	public class SimpleInjectorBootstrapProvider : IBootstrapProvider
	{
		Container _container;

		public Task Init(string name, IProviderRuntime providerRuntime, IProviderConfiguration config)
		{
			// Create Kernel
			_container = KernelFactory.Create(new ExecutionContextScopeLifestyle(), true);

			// Prepare Scope
			DependencyResolverContext.OnBeforeInvokeGrainMethod = OnBeforeInvokeGrainMethod;

			return TaskDone.Done;
		}

		IDisposable OnBeforeInvokeGrainMethod(Grain obj)
		{
			var scope = _container.BeginExecutionContextScope();

			// Find Properties with Inject Attribute
			var attribs = obj.GetType().GetProperties().Where(x => x.GetCustomAttributes<InjectAttribute>(true).Any());
			foreach (var attrib in attribs) {
				// Inject Property
				var propertyType = attrib.PropertyType;
				var injectedValue = _container.GetInstance(propertyType);
				attrib.SetValue(obj, injectedValue);
			}

			return scope;
		}

		public string Name
		{
			get { return "SimpleInjectorBootstrapProvider"; }
		}
	}
````

It's not perfect, but the intention of the prototype is to:
1. Allow a smaller amount of injected dependencies be alive at any one moment (grains that aren't actively processing don't need a valid reference to a dependency - waste of memory)
1. Ideally these dependencies were pooled between Orleans worker threads so that only a handful were ever instantiated and we can keep updating the references between each Grain method invocation

I'm sure its horribly broken though with regards to timers etc.

I think a "one time" setup of dependencies (when you could have 100k+ grains) isn't the best avenue if they aren't all Singletons/shared (so constructor or OnActivateAsync overloads would be a one time setup), where-as if they were property based we could keep switching out the dependencies injected based on which worker thread is actively processing a message - my rough thoughts, as opposed to turning all dependencies into Singleton's to keep them memory efficient.

**Comment by nehmebilal** [comment]
Looks good but injecting dependencies at each message might be problematic.

Let's say `HelloGrain` have a dependency on `IWeatherService` which is not thread-safe and is statefull. I can configure my DI container to create a new instance of `WeatherServiceImpl` every time it's asked for one (because it's not thread-safe and multiple grain may be using it in parallel). If the dependencies are injected on activation of a grain, the grain will get the new instance and can safely assume that it's state will stay in memory across messages as long as the grain is active. I don't see how this can be done if the dependencies were injected at each call to the grain.

Injecting on activation is natural for an actor as injecting on construction is natural for an object.

**Comment by attilah** [comment]
Since this topic raised again yesterday night I have finished my implementation of Orleans DI.

Later today I will push this to my fork, so you can take a look.

Since I am a fan of Autofac I made an adapter for that, but any DI container is pluggable.

**Comment by nehmebilal** [comment]
> @attilah Implement other dependency scopes: per grain, ...

What is the current dependency scope?

I haven't read the code yet (I will) but It would be great if you can provide a high level description with code samples.

**Comment by attilah** [comment]
Per silo is the current scope.

**Comment by Plasma** [comment]
This sounds great, a per Grain and Silo scope support would be a good start for me.

**Comment by nehmebilal** [comment]
@attilah I tried to look at your code but it's very difficult to see all the changes you've done regarding DI because you pulled directly from *dotnet/orleans* and did not use `git rebase` to keep your commits together. As a result, your commits are scattered and very difficult to track. Take a look at the following [guide](http://2buntu.com/articles/1459/keeping-your-forked-repo-synced-with-the-upstream-source/) for more information on how to keep your history clean while staying up-to-date with the upstream source. It's also better to create a branch instead of working directly in master.

As far as I can tell, this work doesn't solve the problem discussed in this issue. It does change the way Orleans creates grains by introducing the dependency resolver but I don't see how this allows one to inject dependencies into a grain?

It would be great if you can take a look at the example provided in the description of this issue and tell us how your implementation can be used to inject `IWeatherService` into a grain. Sorry if I am missing something.

**Comment by attilah** [comment]
@nehmebilal sorry for not being a git guru, windiff and other diff tools against the orleans master can make it clear what are my modifications. In the detailed comment above I described EVERY piece I touched, so by looking just at them you can see what was done.

You are right, my modification is adding DI capablities to Orleans itself, a framework like Orleans SHOULD support DI and SHOULD NOT implement it. All those [Inject] samples are putting DI implementation into Orleans itself. That functionality - the injecting - is the role of a DI container, not Orleans.

How is the injection possible? all injection possibilities are dependent on the DI container you choose, lets take the simplest one: ctor injection.

if you're registering a service in your own project dependent DepencyResolverProvider class like this:

```
container.Register<UserService>().As<IUserService>().SingleInstance();
````

and having a grain class defined:

```
public class SimpleGrain : ISimpleGrain
{
    private readonly IUserService userService;

    public SimpleGrain(IUserService userService)
    {
        this.userService = userService;
    }
}
```

then since the configured DI container is responsible to create the grains, the whol DI mechanism will kick in and Autofac in this case will try to lookup the registered IUserService and inject it into the constructor parameter userService.

This pretty much solves the problem of injecting your own class instances into your grains.

Is it more clear now?!

**Comment by nehmebilal** [comment]
> @attilah  @nehmebilal sorry for not being a git guru, windiff and other diff tools against the orleans master can make it clear what are my modifications. In the detailed comment above I described EVERY piece I touched, so by looking just at them you can see what was done.

Sorry if I hurt your feelings :) I was just pointing out the recommended way of using git/github. The article is probably worth reading and makes it easier for others to understand and value your contributions, which I personally greatly appreciate.

> Is it more clear now?!

A lot more clear! It always is when you take the time to explain it :)

I personally like the fact that the grain will automatically support whatever DI mechanism is supported by the DI container (constructor injection, property injection, etc). In other words, no need to manually inspect the grain and figure out what needs to be injected.

On the other hand, I have some thoughts/concerns:

1) The injection mechanism is coupled to the DI container. If I have a grain that uses property injection by marking a property with `[inject]` attribute and I switch to another DI container that uses `[autowire]` instead of `[inject]`, I have to update all my grains code. On the other hand, if all DI containers use standard attributes name (I am not sure if they do in C#?), this won't be an issue.

2) Users can easily do mistakes like the following:

```csharp
public class SimpleGrain : Grain, ISimpleGrain
{
    private readonly string userInfo;

    public SimpleGrain(IUserService userService)
    {
        // the following will blow up because IGrainIdentity has not been set yet
        this.userInfo = userService.GetInfo(this.GetPrimaryKeyString());
    }
}
``` 

3) What about the `GrainFactory` property of a `Grain`? Should we get rid of it since other grains can be injected into `MyGrain`? (DI all the way)

**Comment by attilah** [comment]
@nehmebilal 

Oh it is not about my feelings ;-) I know that I've git limitations at this level, which I'll have to practice ;-)

If you check out the PR: https://github.com/dotnet/orleans/pull/709 I recreated my fork to have 1 commit for the 1 PR :-) it was much more less time now that I spent with the rebase magic :-(

From my experience you choose a DI container technology at the start of the project, and not changing that in between. If you're changing that, yes, you've to refactor. But it is also true for ANY other project types and frameworks where you're using DI containers. So in this aspect your 1st point is not applicable to Orleans I think.

That example you added will blow up at the first test. Since you're writing tests, right? :-)

You can not protect the developer from himself. If we can shoot ourselves in the foot, we will.

Regarding GrainFactory: I'm not sure, but grain instances will not be injectable into your other grains, since development time you don't know (ok, you can for singleton or system grains) the Id, and you've no standard way of specifying a grain Id in your code.

If we'll have something like ActivationScope, which makes sure that the injection will take place before invoking the OnActivating method...maybe it will be possible, but don't run that far away ;-)

**Comment by james-andrewsmith** [comment]
I've been lurking on these DI threads a while, as I'm very keen to see it implemented in Orleans, I agree with @attilah that the DI injection should all be handled by the developers chosen DI framework. 

@nehmebilal My 2 cents on the below.

> 1) The injection mechanism is coupled to the DI container. If I have a grain that uses property injection by marking a property with [inject] attribute and I switch to another DI container that uses [autowire] instead of [inject], I have to update all my grains code. On the other hand, if all DI containers use standard attributes name (I am not sure if they do in C#?), this won't be an issue.

I feel this is the responsibility of the developer when they choose / migrate a DI framework. Some will be easier than others, but ultimately it's not Orleans job to make it easier, it's something for the DI projects to work out :)

> 2) Users can easily do mistakes like the following:

I always make this mistake, but I always quickly find it when I'm debugging, if an interface isn't known by the DI framework then my preference is things should break.... fast.

> 3) What about the GrainFactory property of a Grain? Should we get rid of it since other grains can be injected into MyGrain? (DI all the way)

The way I understand how GrainFactory works, wouldn't this make implementing DI a much more complicated task? Either way, my intended use cases for GrainFactory don't need DI/IOC (Also using DI in place of GrainFactory feels a lot like a ServiceLocator to me).

**Comment by nehmebilal** [comment]
> @attilah If you check out the PR: #709 I recreated my fork to have 1 commit for the 1 PR :-)

Thanks!

> @attilah Regarding GrainFactory: I'm not sure, but grain instances will not be injectable into your other grains, since development time you don't know (ok, you can for singleton or system grains) the Id, and you've no standard way of specifying a grain Id in your code.

What about named dependencies? Here is an example I copied from [here](https://msdn.microsoft.com/en-us/library/ff649447.aspx):
```csharp
public class MyObject
{
  private IMyInterface _objA, _objB;

  [Dependency("MapTypeA")]
  public IMyInterface ObjectA
  {
    get { return _objA; }
    set { _objA = value; }
  }

  [Dependency("MapTypeB")]
  public IMyInterface ObjectB
  {
    get { return _objB; }
    set { _objB = value; }
  }
} 
```

But it's true that in some cases grains ids are generated at runtime (e.g. computation grains) and this is where the `GrainFactory` would still be needed.

In short, great work @attilah! It's time to look at the PR :)

**Comment by jason-bragg** [comment]
There are two areas I feel it’s important to build a consensus around prior to accepting a solution (PR).

1)	How does application logic (primarily grain logic) get access to the resolver?
2)	What is the minimal set of functionality the resolver should support?

In regards to pr #709:
1)	– Application access to resolver.
  a.	The grain does not seem to have access to the resolver.  It seems to rely on everything being passed to it at construction time.  This seems unnecessarily limiting as the grain has no way to use the resolver to obtain other dependencies on demand.
  b.	This relies on the resolver to create grains and encourages logic in the grain constructor.  This is not in-line with existing actor model (at least as I understand it).  At construction time, much of the grain functionality is incomplete, so we discourage adding logic in grain constructors.  We would have prevented custom constructors entirely if there was a simple pattern to do so.  In the existing actor model, the first supported point of control provided is the onActivateAsync.
Due to the above issues, I’d suggest adding the resolver as a property on the Grain base class and the grain developer acquire dependencies as needed, with the onActrivateAsync being the first opportunity to do this.  This protects the integrity of the orleans grain creation and initialization while allowing grain’s a clean safe path to obtain injected deputies.
2)	– Resolver interfaces.
  a.	Resolver interfaces have no way to register implementations.  This means that orleans (using the default resolver) has little to no way to support DI unless an external impl is integrated.  Part of the idea of keeping these interfaces minimal was to provide a minimal set of ‘useful’ DI functionality.  Minimal, so we don't find ourselves re-creating the wheel.  While minimal, these interfaces seem incomplete.
I suggest the first pass of this provide a minimal useful set of interfaces as well as default DI implementation.  After the default minimal capabilities are vetted, then integrating with existing, more robust, third party packages can be introduced.

**Comment by nehmebilal** [comment]
> @jason-bragg 1) – Application access to resolver.
a. The grain does not seem to have access to the resolver. It seems to rely on everything being passed to it at construction time. This seems unnecessarily limiting as the grain has no way to use the resolver to obtain other dependencies on demand.

Why would it need the resolver? The fact that the grain has no access to the resolver makes it easy to tell what are the dependencies of the grain by looking at the constructor / properties. This improves code readability.

> b. This relies on the resolver to create grains and encourages logic in the grain constructor. This is not in-line with existing actor model (at least as I understand it). At construction time, much of the grain functionality is incomplete, so we discourage adding logic in grain constructors. We would have prevented custom constructors entirely if there was a simple pattern to do so. In the existing actor model, the first supported point of control provided is the onActivateAsync.

:+1: This makes perfect sense to me. 

On the other hand, nothing prevent users today from doing the same mistake: they can implement a default constructor in their grain code and add logic to it.

What I really like about #709 approach is that it allows one to use whatever DI mechanism is supported by the DI container he chooses (constructor injection, setter injection, named dependencies, etc.). This seems very powerful and flexible and takes away the need to create and maintain the resolver interfaces in Orleans.

**Comment by jason-bragg** [comment]
@nehmebilal 
"Why would it need the resolver? The fact that the grain has no access to the resolver makes it easy to tell what are the dependencies of the grain by looking at the constructor / properties."
This assumes that all dependencies are known at construction time, but the resolver itself can be passed in through the constructor to address these scenarios, so I withdraw this concern.

"nothing prevent users today from doing the same mistake"
Yes, but if someone does put logic in the constructor and something goes wrong, for instance they try to use the grain factory and it's not yet set, the answer is:
'Don't do that.  Do you're initialization in OnActivateAsync.  That's what that function is for. We don't support logic in the constructor.'
Supporting dependency injection through the constructor is in direct conflict to this.

I'm not arguing we should avoid injecting dependencies in the constructor, I'm saying we need to have a clear story here.  We can keep things as they are, and make a resolver available to users (as a grain property?) that can be used to resolve dependencies in the OnActivateAsync, or we can allow injection of grain class, and dependencies through the constructor, accepting the challenge of ensuring all necessary initialization logic is complete prior to running the grain constructor.  We don't even have to do all this work in the first pass, just have an understanding of the issues and agreement on the direction.

"What I really like about #709 approach is that it allows one to use whatever DI mechanism is supported by the DI container he chooses (constructor injection, setter injection, named dependencies, etc.)."
+1
In my view, a big advantage to introducing these capabilities is to allow the simplification and componentization of the Orleans framework itself, for testability, extensibility, and maintainability reasons.  These DI interfaces are insufficient for those purposes.  I'm willing to accept that I'm expecting more from this than I should at this stage.

**Comment by veikkoeeva** [comment]
@jason-bragg OWIN like startup mechanism has been discussed on a few occassions. I believe @attilah has mentioned it too. Perhaps it is still too early to open an issue about wider Orleans componentization, but perhaps we can gather some ideas on how to do it already. #709 looks like a clean base to build on.

**Comment by attilah** [comment]
The current dependency resolver is accessible thru the Silo.Current.DependencyResolver property if needed.

**Comment by gabikliot** [comment]
DI added to Orleans by @attilah in PR #827.
